crates/ironrdp-acceptor/src/finalization.rs

```diff
@@ -80,44 +82,42 @@ impl Sequence for FinalizationSequence {
     fn step(&mut self, input: &[u8], output: &mut WriteBuf) -> ConnectorResult<Written> {
         let (written, next_state) = match std::mem::take(&mut self.state) {
             FinalizationState::WaitSynchronize => {
-                let data = pdu::decode::<pdu::mcs::SendDataRequest<'_>>(input).map_err(ConnectorError::pdu)?;
-
-                let synchronize = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
+                let synchronize = decode_share_control(input);
 
                 debug!(message = ?synchronize, "Received");
 
                 (Written::Nothing, FinalizationState::WaitControlCooperate)
             }
 
             FinalizationState::WaitControlCooperate => {
-                let data = pdu::decode::<pdu::mcs::SendDataRequest<'_>>(input).map_err(ConnectorError::pdu)?;
-
-                let cooperate = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
+                let cooperate = decode_share_control(input);
 
                 debug!(message = ?cooperate, "Received");
 
                 (Written::Nothing, FinalizationState::WaitRequestControl)
             }
 
             FinalizationState::WaitRequestControl => {
-                let data = pdu::decode::<pdu::mcs::SendDataRequest<'_>>(input).map_err(ConnectorError::pdu)?;
-
-                let control = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
+                let control = decode_share_control(input)?;
 
                 debug!(message = ?control, "Received");
 
                 (Written::Nothing, FinalizationState::WaitFontList)
             }
 
-            FinalizationState::WaitFontList => {
-                let data = pdu::decode::<pdu::mcs::SendDataRequest<'_>>(input).map_err(ConnectorError::pdu)?;
+            FinalizationState::WaitFontList => match decode_font_list(input) {
+                Ok(font_list) => {
+                    debug!(message = ?font_list, "Received");
 
-                let font_list = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
+                    (Written::Nothing, FinalizationState::SendSynchronizeConfirm)
+                }
 
-                debug!(message = ?font_list, "Received");
+                Err(input_event) => {
+                    self.input_events.push(input_event);
```

Comment:
May be worth to add a debug or trace log, just in case this could be relevant for debugging something